File: 2019/08/rs/src/main.rs
use num::complex::Complex;
use std::collections::HashMap;

const IMAGE_WIDTH: u32 = 25;
const IMAGE_HEIGHT: u32 = 6;
const CHARACTER_WIDTH: u32 = 5;
const PIXELS_PER_LAYER: u32 = IMAGE_WIDTH * IMAGE_HEIGHT;

struct LetterMap {
    letters: HashMap<u32, char>,
}

impl LetterMap {
    fn new() -> LetterMap {
        LetterMap {
            letters: vec![
                (
                      (0b01100 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b10010 << CHARACTER_WIDTH * 2)
                    + (0b11110 << CHARACTER_WIDTH * 3)
                    + (0b10010 << CHARACTER_WIDTH * 4)
                    + (0b10010 << CHARACTER_WIDTH * 5),
                    'A',
                ),
                (
                      (0b11100 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b11100 << CHARACTER_WIDTH * 2)
                    + (0b10010 << CHARACTER_WIDTH * 3)
                    + (0b10010 << CHARACTER_WIDTH * 4)
                    + (0b11100 << CHARACTER_WIDTH * 5),
                    'B',
                ),
                (
                      (0b01100 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b10000 << CHARACTER_WIDTH * 2)
                    + (0b10000 << CHARACTER_WIDTH * 3)
                    + (0b10010 << CHARACTER_WIDTH * 4)
                    + (0b01100 << CHARACTER_WIDTH * 5),
                    'C',
                ),
                (
                      (0b11100 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b10010 << CHARACTER_WIDTH * 2)
                    + (0b10010 << CHARACTER_WIDTH * 3)
                    + (0b10010 << CHARACTER_WIDTH * 4)
                    + (0b11100 << CHARACTER_WIDTH * 5),
                    'D',
                ),
                (
                      (0b11110 << CHARACTER_WIDTH * 0)
                    + (0b10000 << CHARACTER_WIDTH * 1)
                    + (0b11100 << CHARACTER_WIDTH * 2)
                    + (0b10000 << CHARACTER_WIDTH * 3)
                    + (0b10000 << CHARACTER_WIDTH * 4)
                    + (0b11110 << CHARACTER_WIDTH * 5),
                    'E',
                ),
                (
                      (0b11110 << CHARACTER_WIDTH * 0)
                    + (0b10000 << CHARACTER_WIDTH * 1)
                    + (0b11100 << CHARACTER_WIDTH * 2)
                    + (0b10000 << CHARACTER_WIDTH * 3)
                    + (0b10000 << CHARACTER_WIDTH * 4)
                    + (0b10000 << CHARACTER_WIDTH * 5),
                    'F',
                ),
                (
                      (0b01100 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b10000 << CHARACTER_WIDTH * 2)
                    + (0b10110 << CHARACTER_WIDTH * 3)
                    + (0b10010 << CHARACTER_WIDTH * 4)
                    + (0b01110 << CHARACTER_WIDTH * 5),
                    'G',
                ),
                (
                      (0b10010 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b11110 << CHARACTER_WIDTH * 2)
                    + (0b10010 << CHARACTER_WIDTH * 3)
                    + (0b10010 << CHARACTER_WIDTH * 4)
                    + (0b10010 << CHARACTER_WIDTH * 5),
                    'H',
                ),
                (
                      (0b01110 << CHARACTER_WIDTH * 0)
                    + (0b00100 << CHARACTER_WIDTH * 1)
                    + (0b00100 << CHARACTER_WIDTH * 2)
                    + (0b00100 << CHARACTER_WIDTH * 3)
                    + (0b00100 << CHARACTER_WIDTH * 4)
                    + (0b01110 << CHARACTER_WIDTH * 5),
                    'I',
                ),
                (
                      (0b00110 << CHARACTER_WIDTH * 0)
                    + (0b00010 << CHARACTER_WIDTH * 1)
                    + (0b00010 << CHARACTER_WIDTH * 2)
                    + (0b00010 << CHARACTER_WIDTH * 3)
                    + (0b10010 << CHARACTER_WIDTH * 4)
                    + (0b01100 << CHARACTER_WIDTH * 5),
                    'J',
                ),
                (
                      (0b10010 << CHARACTER_WIDTH * 0)
                    + (0b10100 << CHARACTER_WIDTH * 1)
                    + (0b11000 << CHARACTER_WIDTH * 2)
                    + (0b10100 << CHARACTER_WIDTH * 3)
                    + (0b10100 << CHARACTER_WIDTH * 4)
                    + (0b10010 << CHARACTER_WIDTH * 5),
                    'K',
                ),
                (
                      (0b10000 << CHARACTER_WIDTH * 0)
                    + (0b10000 << CHARACTER_WIDTH * 1)
                    + (0b10000 << CHARACTER_WIDTH * 2)
                    + (0b10000 << CHARACTER_WIDTH * 3)
                    + (0b10000 << CHARACTER_WIDTH * 4)
                    + (0b11110 << CHARACTER_WIDTH * 5),
                    'L',
                ),
                (
                      (0b01100 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b10010 << CHARACTER_WIDTH * 2)
                    + (0b10010 << CHARACTER_WIDTH * 3)
                    + (0b10010 << CHARACTER_WIDTH * 4)
                    + (0b01100 << CHARACTER_WIDTH * 5),
                    'O',
                ),
                (
                      (0b11100 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b10010 << CHARACTER_WIDTH * 2)
                    + (0b11100 << CHARACTER_WIDTH * 3)
                    + (0b10000 << CHARACTER_WIDTH * 4)
                    + (0b10000 << CHARACTER_WIDTH * 5),
                    'P',
                ),
                (
                      (0b11100 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b10010 << CHARACTER_WIDTH * 2)
                    + (0b11100 << CHARACTER_WIDTH * 3)
                    + (0b10100 << CHARACTER_WIDTH * 4)
                    + (0b10010 << CHARACTER_WIDTH * 5),
                    'R',
                ),
                (
                      (0b01110 << CHARACTER_WIDTH * 0)
                    + (0b10000 << CHARACTER_WIDTH * 1)
                    + (0b10000 << CHARACTER_WIDTH * 2)
                    + (0b01100 << CHARACTER_WIDTH * 3)
                    + (0b00010 << CHARACTER_WIDTH * 4)
                    + (0b11100 << CHARACTER_WIDTH * 5),
                    'S',
                ),
                (
                      (0b10010 << CHARACTER_WIDTH * 0)
                    + (0b10010 << CHARACTER_WIDTH * 1)
                    + (0b10010 << CHARACTER_WIDTH * 2)
                    + (0b10010 << CHARACTER_WIDTH * 3)
                    + (0b10010 << CHARACTER_WIDTH * 4)
                    + (0b01100 << CHARACTER_WIDTH * 5),
                    'U',
                ),
                (
                      (0b10001 << CHARACTER_WIDTH * 0)
                    + (0b10001 << CHARACTER_WIDTH * 1)
                    + (0b01010 << CHARACTER_WIDTH * 2)
                    + (0b00100 << CHARACTER_WIDTH * 3)
                    + (0b00100 << CHARACTER_WIDTH * 4)
                    + (0b00100 << CHARACTER_WIDTH * 5),
                    'Y',
                ),
                (
                      (0b11110 << CHARACTER_WIDTH * 0)
                    + (0b00010 << CHARACTER_WIDTH * 1)
                    + (0b00100 << CHARACTER_WIDTH * 2)
                    + (0b01000 << CHARACTER_WIDTH * 3)
                    + (0b10000 << CHARACTER_WIDTH * 4)
                    + (0b11110 << CHARACTER_WIDTH * 5),
                    'Z',
                ),
            ]
            .into_iter()
            .collect::<HashMap<u32, char>>(),
        }
    }
    fn get(&self, code: u32) -> char {
        *self.letters.get(&code).unwrap()
    }
}

fn get_image_layers(pixels: &Vec<u32>) -> Vec<Vec<u32>> {
    let pixels_per_layer = PIXELS_PER_LAYER as usize;
    let layer_count = pixels.len() / pixels_per_layer;
    (0..layer_count)
        .into_iter()
        .map(|layer_index| {
            pixels[(layer_index * pixels_per_layer)..(pixels_per_layer * (layer_index + 1))]
                .iter()
                .map(|pixel| *pixel)
                .collect()
        })
        .collect()
}

fn part1(layers: &Vec<Vec<u32>>) -> usize {
    let mut least_zeros = usize::MAX;
    let mut result = 0;
    for layer in layers {
        let zero_count = layer.iter().filter(|pixel| **pixel == 0).count();
        if zero_count < least_zeros {
            least_zeros = zero_count;
            result = layer.iter().filter(|pixel| **pixel == 1).count()
                * layer.iter().filter(|pixel| **pixel == 2).count()
        }
    }
    result
}

const _BLACK: u32 = 0;
const WHITE: u32 = 1;
const TRANSPARENT: u32 = 2;

fn get_character_in_screen(
    screen: &HashMap<Complex<u32>, u32>,
    index: u32,
    width: u32,
    height: u32,
    letter_map: &LetterMap,
) -> char {
    let mut screen_value = 0;
    for x in 0..width {
        for y in 0..height {
            if screen.get(&Complex::new(width * index + x, y)).unwrap() == &WHITE {
                screen_value += u32::pow(2, width - 1 - x) << (y * width);
            }
        }
    }
    letter_map.get(screen_value)
}

fn part2(layers: &Vec<Vec<u32>>) -> String {
    let letter_map = LetterMap::new();
    let mut image = (0..(IMAGE_WIDTH * IMAGE_HEIGHT))
        .into_iter()
        .map(|index| {
            (
                Complex::new(index % IMAGE_WIDTH, index / IMAGE_WIDTH),
                TRANSPARENT,
            )
        })
        .collect::<HashMap<Complex<u32>, u32>>();
    for layer in layers {
        for x in 0..IMAGE_WIDTH {
            for y in 0..IMAGE_HEIGHT {
                let pixel = Complex::new(x, y);
                if image.get(&pixel).unwrap() < &2 {
                    continue;
                }
                *image.get_mut(&pixel).unwrap() = layer[(x + y * IMAGE_WIDTH) as usize];
            }
        }
    }
    (0..(IMAGE_WIDTH / CHARACTER_WIDTH))
        .into_iter()
        .map(|index| {
            get_character_in_screen(&image, index, CHARACTER_WIDTH, IMAGE_HEIGHT, &letter_map)
        })
        .collect::<String>()
}

fn solve(pixels: &Vec<u32>) -> (usize, String) {
    let layers = get_image_layers(pixels);
    (part1(&layers), part2(&layers))
}

fn get_input(file_path: &String) -> Vec<u32> {
    std::fs::read_to_string(file_path)
        .expect("Error reading input file!")
        .chars()
        .map(|i| i.to_digit(10).unwrap())
        .collect()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    let end = now.elapsed().as_secs_f32();
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:.7}", end);
}

File: 2019/08/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <math.h>

typedef struct {
    int count, *pixels;
} Input;
typedef struct
{
    int part1;
    char *part2;
} Results;

#define IMAGE_WIDTH 25
#define IMAGE_HEIGHT 6
#define PIXELS_PER_LAYER (IMAGE_WIDTH * IMAGE_HEIGHT)
#define CHARACTER_WIDTH 5

int part1(Input input)
{
    int layerCount = input.count / PIXELS_PER_LAYER;
    int leastZeros = PIXELS_PER_LAYER, zeros, ones, twos, index, result;
    for (int layerIndex = 0; layerIndex < layerCount; layerIndex++)
    {
        zeros = ones = twos = 0;
        for (index = 0; index < PIXELS_PER_LAYER; index++)
            switch (input.pixels[layerIndex * PIXELS_PER_LAYER + index])
            {
                case 0: zeros++; break;
                case 1: ones++; break;
                case 2: twos++; break;
            }
        if (zeros < leastZeros)
        {
            leastZeros = zeros;
            result = ones * twos;
        }
    }
    return result;
}

#define BLACK 0
#define WHITE 1
#define TRANSPARENT 2

#define A \
    (0b01100 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b10010 << CHARACTER_WIDTH * 2) + \
    (0b11110 << CHARACTER_WIDTH * 3) + \
    (0b10010 << CHARACTER_WIDTH * 4) + \
    (0b10010 << CHARACTER_WIDTH * 5)
#define B \
    (0b11100 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b11100 << CHARACTER_WIDTH * 2) + \
    (0b10010 << CHARACTER_WIDTH * 3) + \
    (0b10010 << CHARACTER_WIDTH * 4) + \
    (0b11100 << CHARACTER_WIDTH * 5)
#define C \
    (0b01100 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b10000 << CHARACTER_WIDTH * 2) + \
    (0b10000 << CHARACTER_WIDTH * 3) + \
    (0b10010 << CHARACTER_WIDTH * 4) + \
    (0b01100 << CHARACTER_WIDTH * 5)
#define D \
    (0b11100 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b10010 << CHARACTER_WIDTH * 2) + \
    (0b10010 << CHARACTER_WIDTH * 3) + \
    (0b10010 << CHARACTER_WIDTH * 4) + \
    (0b11100 << CHARACTER_WIDTH * 5)
#define E \
    (0b11110 << CHARACTER_WIDTH * 0) + \
    (0b10000 << CHARACTER_WIDTH * 1) + \
    (0b11100 << CHARACTER_WIDTH * 2) + \
    (0b10000 << CHARACTER_WIDTH * 3) + \
    (0b10000 << CHARACTER_WIDTH * 4) + \
    (0b11110 << CHARACTER_WIDTH * 5)
#define F \
    (0b11110 << CHARACTER_WIDTH * 0) + \
    (0b10000 << CHARACTER_WIDTH * 1) + \
    (0b11100 << CHARACTER_WIDTH * 2) + \
    (0b10000 << CHARACTER_WIDTH * 3) + \
    (0b10000 << CHARACTER_WIDTH * 4) + \
    (0b10000 << CHARACTER_WIDTH * 5)
#define G \
    (0b01100 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b10000 << CHARACTER_WIDTH * 2) + \
    (0b10110 << CHARACTER_WIDTH * 3) + \
    (0b10010 << CHARACTER_WIDTH * 4) + \
    (0b01110 << CHARACTER_WIDTH * 5)
#define H \
    (0b10010 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b11110 << CHARACTER_WIDTH * 2) + \
    (0b10010 << CHARACTER_WIDTH * 3) + \
    (0b10010 << CHARACTER_WIDTH * 4) + \
    (0b10010 << CHARACTER_WIDTH * 5)
#define I \
    (0b01110 << CHARACTER_WIDTH * 0) + \
    (0b00100 << CHARACTER_WIDTH * 1) + \
    (0b00100 << CHARACTER_WIDTH * 2) + \
    (0b00100 << CHARACTER_WIDTH * 3) + \
    (0b00100 << CHARACTER_WIDTH * 4) + \
    (0b01110 << CHARACTER_WIDTH * 5)
#define J \
    (0b00110 << CHARACTER_WIDTH * 0) + \
    (0b00010 << CHARACTER_WIDTH * 1) + \
    (0b00010 << CHARACTER_WIDTH * 2) + \
    (0b00010 << CHARACTER_WIDTH * 3) + \
    (0b10010 << CHARACTER_WIDTH * 4) + \
    (0b01100 << CHARACTER_WIDTH * 5)
#define K \
    (0b10010 << CHARACTER_WIDTH * 0) + \
    (0b10100 << CHARACTER_WIDTH * 1) + \
    (0b11000 << CHARACTER_WIDTH * 2) + \
    (0b10100 << CHARACTER_WIDTH * 3) + \
    (0b10100 << CHARACTER_WIDTH * 4) + \
    (0b10010 << CHARACTER_WIDTH * 5)
#define L \
    (0b10000 << CHARACTER_WIDTH * 0) + \
    (0b10000 << CHARACTER_WIDTH * 1) + \
    (0b10000 << CHARACTER_WIDTH * 2) + \
    (0b10000 << CHARACTER_WIDTH * 3) + \
    (0b10000 << CHARACTER_WIDTH * 4) + \
    (0b11110 << CHARACTER_WIDTH * 5)
#define M 0
#define N 0
#define O \
    (0b01100 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b10010 << CHARACTER_WIDTH * 2) + \
    (0b10010 << CHARACTER_WIDTH * 3) + \
    (0b10010 << CHARACTER_WIDTH * 4) + \
    (0b01100 << CHARACTER_WIDTH * 5)
#define P \
    (0b11100 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b10010 << CHARACTER_WIDTH * 2) + \
    (0b11100 << CHARACTER_WIDTH * 3) + \
    (0b10000 << CHARACTER_WIDTH * 4) + \
    (0b10000 << CHARACTER_WIDTH * 5)
#define Q 0
#define R \
    (0b11100 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b10010 << CHARACTER_WIDTH * 2) + \
    (0b11100 << CHARACTER_WIDTH * 3) + \
    (0b10100 << CHARACTER_WIDTH * 4) + \
    (0b10010 << CHARACTER_WIDTH * 5)
#define S \
    (0b01110 << CHARACTER_WIDTH * 0) + \
    (0b10000 << CHARACTER_WIDTH * 1) + \
    (0b10000 << CHARACTER_WIDTH * 2) + \
    (0b01100 << CHARACTER_WIDTH * 3) + \
    (0b00010 << CHARACTER_WIDTH * 4) + \
    (0b11100 << CHARACTER_WIDTH * 5)
#define T 0
#define U \
    (0b10010 << CHARACTER_WIDTH * 0) + \
    (0b10010 << CHARACTER_WIDTH * 1) + \
    (0b10010 << CHARACTER_WIDTH * 2) + \
    (0b10010 << CHARACTER_WIDTH * 3) + \
    (0b10010 << CHARACTER_WIDTH * 4) + \
    (0b01100 << CHARACTER_WIDTH * 5)
#define V 0
#define W 0
#define X 0
#define Y \
    (0b10001 << CHARACTER_WIDTH * 0) + \
    (0b10001 << CHARACTER_WIDTH * 1) + \
    (0b01010 << CHARACTER_WIDTH * 2) + \
    (0b00100 << CHARACTER_WIDTH * 3) + \
    (0b00100 << CHARACTER_WIDTH * 4) + \
    (0b00100 << CHARACTER_WIDTH * 5)
#define Z \
    (0b11110 << CHARACTER_WIDTH * 0) + \
    (0b00010 << CHARACTER_WIDTH * 1) + \
    (0b00100 << CHARACTER_WIDTH * 2) + \
    (0b01000 << CHARACTER_WIDTH * 3) + \
    (0b10000 << CHARACTER_WIDTH * 4) + \
    (0b11110 << CHARACTER_WIDTH * 5)
const int LETTERS[] = { A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z };
char getLetter(int value)
{
    for (int index = 0; index < 26; index++)
        if (LETTERS[index] == value)
            return 'A' + index;
    perror("Unknow character");
    exit(1);
}

void printImage(int *image)
{
    for (int y = 0; y < IMAGE_HEIGHT; y++)
    {
        for (int x = 0; x < IMAGE_WIDTH; x++)
            printf("%c", image[x + IMAGE_WIDTH * y] == WHITE ? '#' : '.');
        printf("\n");
    }
    printf("\n");
}

char *part2(Input input)
{
    int index, x, y, imageValue;
    int layerCount = input.count / PIXELS_PER_LAYER;
    int *image = malloc(PIXELS_PER_LAYER * sizeof(int));
    for (index = 0; index < PIXELS_PER_LAYER; index++)
        image[index] = TRANSPARENT;
    for (int layerIndex = 0; layerIndex < layerCount; layerIndex++)
        for (index = 0; index < PIXELS_PER_LAYER; index++)
        {
            if (image[index] < TRANSPARENT)
                continue;
            image[index] = input.pixels[layerIndex * PIXELS_PER_LAYER + index];
        }
    char *message = calloc((IMAGE_WIDTH / CHARACTER_WIDTH) + 1, 1);
    for (index = 0; index < IMAGE_WIDTH / CHARACTER_WIDTH; index++)
    {
        imageValue = 0;
        for (x = 0; x < CHARACTER_WIDTH; x++)
            for (y = 0; y < IMAGE_HEIGHT; y++)
                imageValue += (image[CHARACTER_WIDTH * index + x + IMAGE_WIDTH * y] == WHITE) * ((int)pow(2, CHARACTER_WIDTH - 1 - x) << (y * CHARACTER_WIDTH));
        message[index] = getLetter(imageValue);
    }
    return message;
}

Results solve(Input input)
{
    return (Results){part1(input), part2(input)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    Input input = {
        (int)length - 1,
        malloc(length * sizeof(int))
    };
    char c;
    int *pixels = input.pixels;
    while ((c = fgetc(file)) != EOF)
        *(pixels++) = c - '0';
    fclose(file);
    return input;
}

void freeInput(Input input)
{
    free(input.pixels);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %s\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
